worker/raft/raftbackstop/worker.go

```diff
@@ -0,0 +1,257 @@
+	if config.Tag == nil {
+		return errors.NotValidf("nil Tag")
+	}
+	return nil
+}
+
+// NewWorker returns a worker responsible for recovering the raft
+// cluster when it's been reduced to one server that can't become
+// leader.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := &Worker{
+		config:        config,
+		serverDetails: make(chan apiserver.Details),
+	}
+	// Subscribe to API server address changes.
+	unsubscribe, err := config.Hub.Subscribe(
+		apiserver.DetailsTopic,
+		w.apiserverDetailsChanged,
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "subscribing to apiserver details")
+	}
+	// Now that we're subscribed, request the current API server details.
+	req := apiserver.DetailsRequest{
+		Requester: "raft-backstop",
+		LocalOnly: true,
+	}
+	if _, err := config.Hub.Publish(apiserver.DetailsRequestTopic, req); err != nil {
+		return nil, errors.Annotate(err, "requesting current apiserver details")
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			defer unsubscribe()
+			return w.loop()
+		},
+	}); err != nil {
+		unsubscribe()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+type Worker struct {
+	catacomb catacomb.Catacomb
+	config   Config
+
+	serverDetails chan apiserver.Details
+	configUpdated bool
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *Worker) loop() error {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case details := <-w.serverDetails:
+			err := w.maybeRecoverCluster(details)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (w *Worker) maybeRecoverCluster(details apiserver.Details) error {
+	if w.configUpdated {
+		w.config.Logger.Infof("raft configuration already updated; waiting for raft worker restart")
+		return nil
+	}
+	if len(details.Servers) != 1 {
+		return nil
+	}
+	if _, found := details.Servers[w.config.Tag.Id()]; !found {
+		return nil
+	}
+	if w.config.Raft.State() == raft.Leader {
+		return nil
+	}
+
+	raftServers, err := w.getConfiguration()
+	if err != nil {
+		return errors.Annotate(err, "getting raft configuration")
+	}
+
+	numServers := len(raftServers)
+	localServer := raftServers[raft.ServerID(w.config.Tag.String())]
+	if localServer == nil {
+		return nil
+	}
+
+	localVote := localServer.Suffrage != raft.Nonvoter
+	if numServers == 1 && localVote {
+		// The server can vote and has quorum, so it can become leader.
+		return nil
+	}
+
+	err = w.recoverCluster(localServer)
+	return errors.Annotate(err, "recovering cluster")
+}
+
+func (w *Worker) recoverCluster(server *raft.Server) error {
+	// Make a configuration that can be written into the log, and append it.
+	newServer := *server
```

Comment:
As a safety measure, should there be a check and err return for nil server

Reply:
Yup, I'll add this.